### DISCUSSION TITLE
MacOS Deployment to the Mac App Store

### DISCUSSION THREAD

**Discussion by HEIC-to-JPEG-Dev** [AvaloniaUI/Avalonia#15567] [General]
I'm looking at how to publish an App to the Mac App Store and the first sentance on the deployment guide states that "some extra legwork has to be done after your application has gone through the **publishing process**" - and no-where in the documentation is "publishing process" mentioned. 

Is that something basic I'm miss-reading ? or is there some mysterious "process" you have to be in a club to know about.

**Comment by BAndysc** [comment]
Isn't it just `dotnet publish`? :) On Linux and Windows `dotnet publish` is pretty much enough to release the app, but it's not the case on macOS

> ### Manual
> First, publish your application ([dotnet publish documentation](https://docs.microsoft.com/en-us/dotnet/core/tools/dotnet-publish)):
> ```
> dotnet publish -r osx-x64 --configuration Release -p:UseAppHost=true
> ```

**Reply by HEIC-to-JPEG-Dev** [reply]
The fact that you don’t know, says everything about the documentation. Documentation should be concise, detail the points, not use TLA’s until expanded, etc. maybe it is, maybe not, who knows? And why do I need the command line?

**Reply by timunie** [reply]
@HEIC-to-JPEG-Dev contribution to the docs are very welcome. That macos is harder to publish ia no Avalonia only issue btw.

**Reply by HEIC-to-JPEG-Dev** [reply]
I’d love to. But at the moment, I have no idea what to do. Once I figure it out, I’ll give my updates

**Reply by nicu-dev** [reply]
https://avaloniaui.net/blog/the-definitive-guide-to-building-and-deploying-avalonia-applications-for-macos

**Reply by HEIC-to-JPEG-Dev** [reply]
This is a great start, but by no means definiitive or comprehensive. It assumes a lot of knowledge of Apple back-end systems, and Avalonia subtleties. 

For example, "change the project file with..." - so does that affect the Windows compilation ? what is the impact. I'm using the exact same code on Mac and Windows (from a file share for example) - will these changes create an issue. I already know that if I create a project in Windows, it won't actually run on Mac correctly; I have to create the project on a Mac, only then will it work on Mac & Windows. These are the exasperations we face in the field and while we learn and go "ah ha !" - those like myself, just give up. 

Then I see "dotnet publish -r osx-x64 -c Release" - where do you run this, what directory, is this on the Mac or on Windows, can it be run on both, etc. Every stage in the blog assumes things; if I knew those things to begin with, I wouldn't need the blog.

To develop an Avalonia app in Windows I have found to be easy, but when I tried to get the app to work & be published on Mac, I spend weeks trying to figure it out. Eventually gave up and just wrote the app in xCode.

I think the only way I would use Avalonia for Mac/Windows development is if there was a "mac app" that makes all the changes for you, reconfigures the project, signs it, asks for the certificates (and details how to get them) etc.

As I've said before, the more detailed documentation in the Avalonia docs is out of date and wrong while this blog is more of an overview than a detailed way to get an app published on a Mac.

**Reply by nicu-dev** [reply]
`Definitive` is an arguable term in itself, but it's the best we've got yet.

`"change the project file with..."` - those properties are good to have anyway, even for windows compilation

`Then I see "dotnet publish -r osx-x64 -c Release" - where do you run this,` - in the terminal, the the project location (where the .csproj file is)

 `"mac app" that makes all the changes for you, reconfigures the project` - look out f[or this project](https://github.com/quamotion/dotnet-packaging), it will soon be available to package macOS installers automatically

`more detailed documentation` - documentation can always be better indeed, but it is what it is. maybe avalonia is not what you need for now at this point, and will be a better fit as it matures. till then, maybe a web app is a better approach if you need cross-platformity.

One important thing that this article doesn't mention, however, is that to be able to sign and certify your macOS app you need to pay $99 for the Apple developer certificate(apple appleing)

**Reply by MikeCodesDotNET** [reply]
The guide was intended to focus specifically on Apple-related aspects of packaging .NET apps for macOS rather than covering general .NET concepts. However, I can see how assumptions about readers' familiarity with basic .NET concepts might have made it less accessible for some.

Regarding your specific points:

The project file changes mentioned are not Mac-specific. They are [general .NET publishing options](https://learn.microsoft.com/en-us/dotnet/core/deploying/single-file/overview?tabs=cli) that can be applied to builds for any supported platform, including Windows and macOS. These settings can be included in the project file for consistent builds or applied directly in the CLI command. Here's an example command that incorporates those settings:

```
dotnet publish -r osx-x64 -c Release \
  --self-contained true \
  -p:PublishSingleFile=true \
  -p:PublishTrimmed=true \
  -p:PublishReadyToRun=true
```

The only Mac-specific part here is the ```'-r osx-x64'``` flag, which specifies the target runtime. For Windows, you'd use something like ```'-r win-x64'``` instead.

The ```'dotnet publish'``` command is typically run in the project's root directory on either Mac or Windows. This is a [standard .NET practice](https://learn.microsoft.com/en-us/dotnet/core/tools/dotnet-publish) across platforms.

I appreciate your feedback and will try to find some time to update the post to include additional context to make it more accessible to developers with varying levels of experience with .NET.

**Reply by HEIC-to-JPEG-Dev** [reply]
I appreciate the work, really I do. And I can handle the "quirks and workarounds" of AvaloniaUI apps (I would use it) - Deploying to a Mac app store is my ONLY issue. I'm not a Mac user; I do have access to a Mac to compile and test, and I do have a developer account.

I need a KISS approach. Assume nothing, for example, saying "in the terminal" could mean Windows or Mac - that's the level I/we need.

The items I mentioned above were two parts of the blog that assumes, basically everything in it does, for example, creating the pList file assumes this won't be going into the app store as it misses out almost all the requirements for that when using Avalonia; getting the certificates, which ones? from where? - there are details in the Avalonia docs, but they're wrong; "Create an app-specific password for your Apple ID" - what? where do I create it, where do I use it? - 

I guess what I would need is a "Okay, you have an AvaloniaUI Windows app published to the Windows App Store - Here's a step by step guide to getting it into the Mac App Store"

I should note that the .Net CLI project does not mention Mac App Store packaging, only sideloading packages.

**Reply by MikeCodesDotNET** [reply]
> I need a KISS approach.

We're working on it. I try my hardest not to talk about things still in the oven, but we're building a solution that makes packing your Avalonia apps incredibly simple. The goal is to require zero platform-specific experience so that Windows devs can deploy on non-Windows platforms and leave the platform packaging idiosyncrasies for us. 

I can't give a timeline for its release, but I will share that it forms part of a broader plan to provide our users with "pro" tools to help folks be more productive with Avalonia.

**Reply by HEIC-to-JPEG-Dev** [reply]
You just made me very happy ! - I don't even mind paying (Pro sounds like a pay option, and I'm an indie, single developer).

Thank you, I'll keep monitoring this thread for that :)

**Comment by workgroupengineering** [comment]
https://docs.avaloniaui.net/docs/deployment/macOS

**Reply by HEIC-to-JPEG-Dev** [reply]
That points to the exact problem I’m talking about

**Reply by MikeCodesDotNET** [reply]
The document is a little long, but it does have everything you need. If you go to the [Manual](https://docs.avaloniaui.net/docs/deployment/macOS#manual) section, it includes information on how to publish.

**Reply by HEIC-to-JPEG-Dev** [reply]
I gave up. I also checked the MAUI documentation, but publishing doing all that manual work is too much hassle. It's easy in xcode so I'll just use that. Thanks for the help though. I'd come back to it if there was a wizard to help, even if the wizard was "click here to access site 
 URL and do this, download file to me and we move on" and "let me create the info.plist file for you".

**Comment by HEIC-to-JPEG-Dev** [comment]
Okay, I've been giving it another try. It's wildly out of date.

I'm at the "Getting Certificates" https://docs.avaloniaui.net/docs/deployment/macOS#getting-certificates

I'll be happy to help updating if I eventually get it working, but...

"go to Xcode > Preferences > Account > Manage Certificates..." should be "go to Xcode > **Settings**> **Accounts** > Manage Certificates..."

So next is to add "Apple Distribution" certificate - yep, that works
Next is add "3rd Party Mac Developer Installer" - that is not an options, those are...
![image](https://github.com/AvaloniaUI/Avalonia/assets/32410442/36cc0203-dbdd-4d62-8b5d-7431899c1e68)

Which one should I create ?

**Reply by HEIC-to-JPEG-Dev** [reply]
I'm also assuming this needs to be imported into the System Keychain, not the iCloud Keychain ?

**Comment by Noemata** [comment]
Using Visual Studio for Mac does simplify things to a degree.  VS will do some of the heavy lifting for you.

**Reply by damonpkuml** [reply]
Visual Studio for Mac has been discontinued.
